### DISCUSSION TITLE
Should typescript-eslint switch its strict config from warnings to errors?

### DISCUSSION THREAD

**Discussion by JoshuaKGoldberg** [eslint/eslint#16512] [Q&A]
https://twitter.com/slicknet/status/1579527957491564544:

> Warnings were designed to aid in transitioning to using new rules. Mark as a warning to let people know this rule will become an error in the future without breaking the build now. It’s intended as a temporary measure.

The `@typescript-eslint/eslint-plugin` `strict` config uses warnings for all its rules. We set it up that way because those strict rules tend to be more opinionated than what most developers want.

What is the recommendation from the ESLint team on this?

I see also #16365 is in-progress. Is this discussion and/or backing _"don't keep warnings long-term"_ preference something that would fit into that project's scope?

**Comment by nzakas** [comment]
This is really up to you. Warnings were intended to be used temporarily. My belief is that using warnings long term results in people tuning them out as just console noise, so I don’t believe you are getting much benefit with long-term warnings.

That said, warnings are a tool in the toolbox that you can use however you’d like. It’s hard to say in the abstract if it’s a good idea or not, so if you’d like to discuss a particular example it might help narrow the discussion.

**Reply by JoshuaKGoldberg** [reply]
> This is really up to you

Hmm, this is what I was hoping to switch away from. It's confusing to at least me -and I think much of the community- that the ESLint docs don't establish a precedent for how to use warnings. I strongly agree with your preference on using them temporarily 😄.

Putting forward: would you accept an issue/PR adding the preference of using warnings to the docs? I.e. adding a warning callout after the list of rule severities on https://eslint.org/docs/latest/user-guide/configuring/rules#configuring-rules:

> We recommend using `"warn"` or `1` only temporarily, while you actively work on addressing the rule's reports. Using warnings long-term often results in developers tuning them out as just console noise.

(is there a better place?)

> a particular example

Backing context on our `strict` config: some rules are generally good but just a bit too opinionated or not-always-clear-what-to-do-on-reports to be put in the recommended configs. We lump both those categories into the `strict` config.

Within the strict rules (as filtered on https://typescript-eslint.io/rules), [`@typescript-eslint/array-type`](https://typescript-eslint.io/rules/array-type) is a good example. We set it as a warning in the `strict` config because it doesn't directly catch bugs - it's just a stylistic preference that a majority of developers prefer.

For a while, we saw many users use warnings as an indication that a rule _should ideally_ be treated but isn't _necessary_ to fix. So we set all `strict` config values to warnings. 😬

**Reply by bmish** [reply]
I'd be supportive of adding a suggestion to the website about when to use `error` vs. `warn`.

Regarding bug vs. stylistic rules, I don't think `error` vs. `warn` is a good way to distinguish these, given the aforementioned limitations of `warn`. What might make more sense is having the `recommended` focus on bugs, optionally a `strict` config that is stricter/more-opinionated in someway, and a separate `stylistic` (or similar name) config focus on formatting. A separate config for formatting that still uses `error` severity allows users enforce formatting if desired while being able to easily swap it out for alternative formatters like prettier. I've seen other linters/plugins follow this pattern, including [ember-template-lint](https://github.com/ember-template-lint/ember-template-lint#presets) which I work on, and ESLint itself has [moved in the direction of](https://github.com/eslint/eslint#stylistic-rule-updates) de-emphasizing or separating out stylistic rules too.

**Reply by JoshuaKGoldberg** [reply]
> separate `stylistic` (or similar name)

Interesting! My first reaction was that I don't want to add _even more_ configs for people to use. My second was that it makes a lot of sense to separate them out. Once https://github.com/typescript-eslint/typescript-eslint/pull/5251 lands we'll really only be increasing from 1 to 2 choices:
* Which baseline config to use: `recommended`, `recommended-requiring-type-checking`, `strict`
* Whether to include `stylistic`

I also like the word _stylistic_ because it's not the same as _formatting_. We aggressively recommend against formatting rules (in our definition, those that would conflict with formatters such as Prettier).

Filed in https://github.com/typescript-eslint/typescript-eslint/issues/5964.

**Reply by nzakas** [reply]
My concern with updating the docs to be more explicit about recommendations for warnings is that I don't want people to think there is only one correct way to decide whether to use a warning or an error. While my original intention was for warnings to be temporary, I don't think that needs to be the case all the time.

I'm doing more Rust programming now, and the compiler shows both errors, which are things that prevent the program from running, and warnings, which are things that aren't what are expected but don't prevent the program from running. If we were to map this onto ESLint, I'd say that warnings could be things that won't cause runtime errors while errors are things you should definitely address (and will block CI).

In either case, I do thing shareable configs should start out with rules at the "error" severity if you believe it's important for people to follow.

Circling back to the docs -- I'm open to a PR so long as we aren't overly prescriptive. I'd rather give a few different scenarios for when warnings are useful rather than nailing down on one.

**Comment by bmish** [comment]
Personally, I would advocate switching from `warn` to `error` in your `strict` config. As mentioned, warnings are easily ignored and thus not that useful other than for temporary/transitional usage. If someone opts-in to using your `strict` config, then presumably they're interested in actually enforcing stricter/more-opinionated rules with actual violations and not just warnings.

**Comment by JoshuaKGoldberg** [comment]
Closing this out: yes, we did in v6. Thanks!

https://typescript-eslint.io/blog/announcing-typescript-eslint-v6/
